File: backend/internal/model/account_erasure_preview.go
// account_erasure_preview.go — Read-only erasure plan an operator reviews before approving.
package model

import "time"

const (
	ErasureActionDelete    = "delete"
	ErasureActionAnonymize = "anonymize"
)

// ErasurePreview lists the application records the automatic erasure would
// change for one request and how each is changed. PlanDigest binds an operator
// approval to exactly this set of records; any later change requires re-review.
type ErasurePreview struct {
	RequestID   int64                      `json:"requestId"`
	GeneratedAt time.Time                  `json:"generatedAt"`
	PlanDigest  string                     `json:"planDigest"`
	Blockers    []string                   `json:"blockers"`
	Tables      []ErasurePreviewTable      `json:"tables"`
	Files       []string                   `json:"files"`
	Unhandled   []AccountDeletionFootprint `json:"unhandled"`
}

// ErasurePreviewTable is one table's affected rows. Rows is a bounded sample;
// Count is the full number of rows the erasure would change.
type ErasurePreviewTable struct {
	Table          string              `json:"table"`
	Action         string              `json:"action"`
	Columns        []string            `json:"columns"`
	MaskedColumns  []string            `json:"maskedColumns"`
	ChangedColumns []string            `json:"changedColumns"`
	Count          int64               `json:"count"`
	Rows           []ErasurePreviewRow `json:"rows"`
}

// ErasurePreviewRow aligns values with ErasurePreviewTable.Columns. After is set
// only for anonymized rows. A nil value is SQL NULL.
type ErasurePreviewRow struct {
	Before []*string `json:"before"`
	After  []*string `json:"after,omitempty"`
	Note   string    `json:"note,omitempty"`
}

File: admin/src/api/accountErasurePreview.ts
// accountErasurePreview — Read-only erasure plan and plan-bound immediate processing.
import { api } from './client';

export interface ErasurePreviewRow {
  before: (string | null)[];
  after?: (string | null)[];
  note?: string;
}
export interface ErasurePreviewTable {
  table: string;
  action: 'delete' | 'anonymize';
  columns: string[];
  maskedColumns: string[];
  changedColumns: string[];
  count: number;
  rows: ErasurePreviewRow[];
}
export interface ErasureUnhandledReference { table: string; column: string; count: number }
export interface ErasurePreview {
  requestId: number;
  generatedAt: string;
  planDigest: string;
  blockers: string[];
  tables: ErasurePreviewTable[];
  files: string[];
  unhandled: ErasureUnhandledReference[];
}

export function fetchErasurePreview(id: number) {
  return api.get<ErasurePreview>(`/api/admin/account-deletions/${id}/preview`);
}
// The server re-reads the records and refuses if they differ from this reviewed plan.
export function expediteReviewedErasure(id: number, reviewedPlanDigest: string) {
  return api.put<void>(`/api/admin/account-deletions/${id}`, { action: 'expedite', reviewedPlanDigest });
}
